README.md
# Webkul TextField With Label
A quick way to create a TextField aligned with label.

## Features
A package designed to make the process of creating and implementing interactive TextField to your application, providing a seamless user experience.

![login_page_ui](https://github.com/Tanya-Webkul/webkul_textfield_with_label/blob/main/login_page_ui.png?raw=true)

## Getting started

Add the latest version of package to your pubspec.yaml (and run`dart pub get`):
```yaml
dependencies:
  webkul_textfield_with_label: ^0.0.5
```

## Usage

```dart
import 'package:webkul_textfield_with_label/webkul_textfield_with_label.dart';
```


After importing the above line in your code, you can use the TextField just like you use any other widget.

With the help of this package, you can specify the Label & can customize it as per you requirement.


```dart
TextFieldWithLabel(labelText:"...//Write the name");
```

**labelText** is the required property.

## Properties

Here are some of the major properties supported by the package...

<!-- TABLE_GENERATE_START -->

| Properties | Description                                                        |
|--|--------------------------------------------------------------------|
| labelText | Specify the label you want to display as heading of your textField |
| labelTextStyle |  Specify the TextStyle for your label                                                       |
| gapBtwLblAndField | Specify the spacing between Label & TextField. By default it is 6.0
| decoration | Specify the decoration for the TextField with InputDecorationTextField class
<!-- TABLE_GENERATE_END -->


With this package, you can use all the properties provided by the TextFormField in Flutter by default.


## Sample Usage
```dart
TextFieldWithLabel(
                    controller: emailController,
                    labelText: "Email",
                    inputDecorationTextField: InputDecorationTextField(
                      prefixIconImage: const Icon(Icons.person),
                      hint: 'Enter Username',
                      filledColor: Colors.white,
                      isEnabled: true,
                      isFilled: true,
                      contentEdgePadding: const EdgeInsets.all(14.0),
                      focusedInputBorder: OutlineInputBorder(
                          borderSide: const BorderSide(color: Colors.white),
                          borderRadius: BorderRadius.circular(8.0)),
                      enabledInputBorder: OutlineInputBorder(
                          borderSide: const BorderSide(color: Colors.white),
                          borderRadius: BorderRadius.circular(8.0)),
                      errorInputBorder: OutlineInputBorder(
                          borderSide: const BorderSide(color: Colors.red),
                          borderRadius: BorderRadius.circular(8.0)),
                    ),
                    validator: (value) {
                      if (!RegExp(
                              r"^[a-zA-Z0-9.a-zA-Z0-9.!#$%&'*+-/=?^_`{|}~]+@[a-zA-Z0-9]+\.[a-zA-Z]+")
                          .hasMatch(value ?? "")) {
                        return "Incorrect Email";
                      }
                    }),
```


See the **example** directory for a complete sample app.




CHANGELOG.md
[0.0.1] - 2023-06-16

* Initial release.

[0.0.2] - 2023-06-16

* Minor Changes to the README.md file

[0.0.3] - 2023-06-16

* Code Improvements

[0.0.4] - 2023-06-16

* Code Improvements & Warnings removed

[0.0.5] - 2023-10-30

* TextField "isRequired" feature added with Label Text

test/webkul_textfield_with_label_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:webkul_textfield_with_label/webkul_textfield_with_label.dart';
void main() {

}


lib/widget/textfield.dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import '../utils/input_decoration.dart';
export 'package:flutter/services.dart' show SmartQuotesType, SmartDashesType;

/// A [FormField] that contains a [TextField].
///
/// This is a convenience widget that wraps a [TextField] widget in a
/// [FormField].
///
/// A [Form] ancestor is not required. The [Form] simply makes it easier to
/// save, reset, or validate multiple fields at once. To use without a [Form],
/// pass a `GlobalKey<FormFieldState>` (see [GlobalKey]) to the constructor and use
/// [GlobalKey.currentState] to save or reset the form field.
///
/// When a [controller] is specified, its [TextEditingController.text]
/// defines the [initialValue]. If this [FormField] is part of a scrolling
/// container that lazily constructs its children, like a [ListView] or a
/// [CustomScrollView], then a [controller] should be specified.
/// The controller's lifetime should be managed by a stateful widget ancestor
/// of the scrolling container.
///
/// If a [controller] is not specified, [initialValue] can be used to give
/// the automatically generated controller an initial value.
///
/// {@macro flutter.material.textfield.wantKeepAlive}
///
/// Remember to call [TextEditingController.dispose] of the [TextEditingController]
/// when it is no longer needed. This will ensure any resources used by the object
/// are discarded.
///
/// By default, `decoration` will apply the [ThemeData.inputDecorationTheme] for
/// the current context to the [InputDecoration], see
/// [InputDecoration.applyDefaults].
///
/// For a documentation about the various parameters, see [TextField].
///
/// {@tool snippet}
///
/// Creates a [LabelTextFormField] with an [InputDecoration] and validator function.
///
/// ![If the user enters valid text, the TextField appears normally without any warnings to the user](https://flutter.github.io/assets-for-api-docs/assets/material/text_form_field.png)
///
/// ![If the user enters invalid text, the error message returned from the validator function is displayed in dark red underneath the input](https://flutter.github.io/assets-for-api-docs/assets/material/text_form_field_error.png)
///
/// ```dart
/// LabelTextFormField(
///   decoration: const InputDecoration(
///     icon: Icon(Icons.person),
///     hintText: 'What do people call you?',
///     labelText: 'Name *',
///   ),
///   onSaved: (String? value) {
///     // This optional block of code can be used to run
///     // code when the user saves the form.
///   },
///   validator: (String? value) {
///     return (value != null && value.contains('@')) ? 'Do not use the @ char.' : null;
///   },
/// )
/// ```
/// {@end-tool}
///
/// {@tool dartpad}
/// This example shows how to move the focus to the next field when the user
/// presses the SPACE key.
///
/// ** See code in examples/api/lib/material/text_form_field/text_form_field.1.dart **
/// {@end-tool}
///
/// See also:
///
///  * <https://material.io/design/components/text-fields.html>
///  * [TextField], which is the underlying text field without the [Form]
///    integration.
///  * [InputDecorator], which shows the labels and other visual elements that
///    surround the actual text editing widget.
///  * Learn how to use a [TextEditingController] in one of our [cookbook recipes](https://flutter.dev/docs/cookbook/forms/text-field-changes#2-use-a-texteditingcontroller).
class LabelTextFormField extends FormField<String> {
  /// Creates a [FormField] that contains a [TextField].
  ///
  /// When a [controller] is specified, [initialValue] must be null (the
  /// default). If [controller] is null, then a [TextEditingController]
  /// will be constructed automatically and its `text` will be initialized
  /// to [initialValue] or the empty string.
  ///
  /// For documentation about the various parameters, see the [TextField] class
  /// and [TextField.new], the constructor.
  LabelTextFormField({
    super.key,
    this.controller,
    String? initialValue,
    FocusNode? focusNode,
    InputDecorationTextField? decoration = const InputDecorationTextField(),
    TextInputType? keyboardType,
    TextCapitalization textCapitalization = TextCapitalization.none,
    TextInputAction? textInputAction,
    TextStyle? style,
    StrutStyle? strutStyle,
    TextDirection? textDirection,
    TextAlign textAlign = TextAlign.start,
    TextAlignVertical? textAlignVertical,
    bool autofocus = false,
    bool readOnly = false,
    bool? showCursor,
    String obscuringCharacter = '•',
    bool obscureText = false,
    bool autocorrect = true,
    SmartDashesType? smartDashesType,
    SmartQuotesType? smartQuotesType,
    bool enableSuggestions = true,
    MaxLengthEnforcement? maxLengthEnforcement,
    int? maxLines = 1,
    int? minLines,
    bool expands = false,
    int? maxLength,
    ValueChanged<String>? onChanged,
    GestureTapCallback? onTap,
    VoidCallback? onEditingComplete,
    ValueChanged<String>? onFieldSubmitted,
    super.onSaved,
    super.validator,
    List<TextInputFormatter>? inputFormatters,
    bool? enabled,
    double cursorWidth = 2.0,
    double? cursorHeight,
    Radius? cursorRadius,
    Color? cursorColor,
    Brightness? keyboardAppearance,
    EdgeInsets scrollPadding = const EdgeInsets.all(20.0),
    bool? enableInteractiveSelection,
    TextSelectionControls? selectionControls,
    InputCounterWidgetBuilder? buildCounter,
    ScrollPhysics? scrollPhysics,
    Iterable<String>? autofillHints,
    AutovalidateMode? autovalidateMode,
    ScrollController? scrollController,
    super.restorationId,
    bool enableIMEPersonalizedLearning = true,
    MouseCursor? mouseCursor,
  })  : assert(initialValue == null || controller == null),
        assert(obscuringCharacter.length == 1),
        assert(maxLines == null || maxLines > 0),
        assert(minLines == null || minLines > 0),
        assert(
          (maxLines == null) || (minLines == null) || (maxLines >= minLines),
          "minLines can't be greater than maxLines",
        ),
        assert(
          !expands || (maxLines == null && minLines == null),
          'minLines and maxLines must be null when expands is true.',
        ),
        assert(!obscureText || maxLines == 1,
            'Obscured fields cannot be multiline.'),
        assert(maxLength == null ||
            maxLength == TextField.noMaxLength ||
            maxLength > 0),
        super(
          initialValue:
              controller != null ? controller.text : (initialValue ?? ''),
          enabled: enabled ?? decoration?.enabled ?? true,
          autovalidateMode: autovalidateMode ?? AutovalidateMode.disabled,
          builder: (FormFieldState<String> field) {
            final _LabelTextFormFieldState state =
                field as _LabelTextFormFieldState;
            final InputDecoration effectiveDecoration = (decoration ??
                    const InputDecoration())
                .applyDefaults(Theme.of(field.context).inputDecorationTheme);
            void onChangedHandler(String value) {
              field.didChange(value);
              if (onChanged != null) {
                onChanged(value);
              }
            }

            return UnmanagedRestorationScope(
              bucket: field.bucket,
              child: TextField(
                restorationId: restorationId,
                controller: state._effectiveController,
                focusNode: focusNode,
                decoration:
                    effectiveDecoration.copyWith(errorText: field.errorText),
                keyboardType: keyboardType,
                textInputAction: textInputAction,
                style: style,
                strutStyle: strutStyle,
                textAlign: textAlign,
                textAlignVertical: textAlignVertical,
                textDirection: textDirection,
                textCapitalization: textCapitalization,
                autofocus: autofocus,
                readOnly: readOnly,
                showCursor: showCursor,
                obscuringCharacter: obscuringCharacter,
                obscureText: obscureText,
                autocorrect: autocorrect,
                smartDashesType: smartDashesType ??
                    (obscureText
                        ? SmartDashesType.disabled
                        : SmartDashesType.enabled),
                smartQuotesType: smartQuotesType ??
                    (obscureText
                        ? SmartQuotesType.disabled
                        : SmartQuotesType.enabled),
                enableSuggestions: enableSuggestions,
                maxLengthEnforcement: maxLengthEnforcement,
                maxLines: maxLines,
                minLines: minLines,
                expands: expands,
                maxLength: maxLength,
                onChanged: onChangedHandler,
                onTap: onTap,
                onEditingComplete: onEditingComplete,
                onSubmitted: onFieldSubmitted,
                inputFormatters: inputFormatters,
                enabled: enabled ?? decoration?.enabled ?? true,
                cursorWidth: cursorWidth,
                cursorHeight: cursorHeight,
                cursorRadius: cursorRadius,
                cursorColor: cursorColor,
                scrollPadding: scrollPadding,
                scrollPhysics: scrollPhysics,
                keyboardAppearance: keyboardAppearance,
                enableInteractiveSelection:
                    enableInteractiveSelection ?? (!obscureText || !readOnly),
                selectionControls: selectionControls,
                buildCounter: buildCounter,
                autofillHints: autofillHints,
                scrollController: scrollController,
                enableIMEPersonalizedLearning: enableIMEPersonalizedLearning,
                mouseCursor: mouseCursor,
              ),
            );
          },
        );

  /// Controls the text being edited.
  ///
  /// If null, this widget will create its own [TextEditingController] and
  /// initialize its [TextEditingController.text] with [initialValue].
  final TextEditingController? controller;

  @override
  FormFieldState<String> createState() => _LabelTextFormFieldState();
}

class _LabelTextFormFieldState extends FormFieldState<String> {
  RestorableTextEditingController? _controller;

  TextEditingController get _effectiveController =>
      _textFormField.controller ?? _controller!.value;

  LabelTextFormField get _textFormField => super.widget as LabelTextFormField;

  @override
  void restoreState(RestorationBucket? oldBucket, bool initialRestore) {
    super.restoreState(oldBucket, initialRestore);
    if (_controller != null) {
      _registerController();
    }
    // Make sure to update the internal [FormFieldState] value to sync up with
    // text editing controller value.
    setValue(_effectiveController.text);
  }

  void _registerController() {
    assert(_controller != null);
    registerForRestoration(_controller!, 'controller');
  }

  void _createLocalController([TextEditingValue? value]) {
    assert(_controller == null);
    _controller = value == null
        ? RestorableTextEditingController()
        : RestorableTextEditingController.fromValue(value);
    if (!restorePending) {
      _registerController();
    }
  }

  @override
  void initState() {
    super.initState();
    if (_textFormField.controller == null) {
      _createLocalController(widget.initialValue != null
          ? TextEditingValue(text: widget.initialValue!)
          : null);
    } else {
      _textFormField.controller!.addListener(_handleControllerChanged);
    }
  }

  @override
  void didUpdateWidget(LabelTextFormField oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (_textFormField.controller != oldWidget.controller) {
      oldWidget.controller?.removeListener(_handleControllerChanged);
      _textFormField.controller?.addListener(_handleControllerChanged);

      if (oldWidget.controller != null && _textFormField.controller == null) {
        _createLocalController(oldWidget.controller!.value);
      }

      if (_textFormField.controller != null) {
        setValue(_textFormField.controller!.text);
        if (oldWidget.controller == null) {
          unregisterFromRestoration(_controller!);
          _controller!.dispose();
          _controller = null;
        }
      }
    }
  }

  @override
  void dispose() {
    _textFormField.controller?.removeListener(_handleControllerChanged);
    _controller?.dispose();
    super.dispose();
  }

  @override
  void didChange(String? value) {
    super.didChange(value);

    if (_effectiveController.text != value) {
      _effectiveController.text = value ?? '';
    }
  }

  @override
  void reset() {
    // setState will be called in the superclass, so even though state is being
    // manipulated, no setState call is needed here.
    _effectiveController.text = widget.initialValue ?? '';
    super.reset();
  }

  void _handleControllerChanged() {
    // Suppress changes that originated from within this class.
    //
    // In the case where a controller has been passed in to this widget, we
    // register this change listener. In these cases, we'll also receive change
    // notifications for changes originating from within this class -- for
    // example, the reset() method. In such cases, the FormField value will
    // already have been set.
    if (_effectiveController.text != value) {
      didChange(_effectiveController.text);
    }
  }
}


lib/utils/input_decoration.dart
import 'package:flutter/material.dart';

class InputDecorationTextField extends InputDecoration {
  final InputBorder? inputBorder;
  final InputBorder? focusedInputBorder;
  final InputBorder? enabledInputBorder;
  final bool? isEnabled;
  final String? errorMsg;
  final InputBorder? errorInputBorder;
  final InputBorder? disabledInputBorder;
  final InputBorder? focusedErrorInputBorder;
  final String? hint;
  final bool? isFilled;
  final Color? filledColor;
  final TextStyle? prefixTextStyling;
  final TextStyle? suffixTextStyling;
  final String? prefixTxt;
  final Widget? prefixIconImage;
  final Color? prefixColor;
  final BoxConstraints? prefixConstraints;
  final String? suffixTxt;
  final Widget? suffixIconImage;
  final Color? suffixColor;
  final BoxConstraints? suffixConstraints;
  final String? helperTxt;
  final TextStyle? helperTxtStyle;
  final int? helperTxtMaxLines;
  final Color? onHoverColor;
  final Color? onFocusColor;
  final TextDirection? hintTxtDirection;
  final TextStyle? hintTxtStyle;
  final int? hintTxtMaxLines;
  final TextStyle? errorTxtStyle;
  final int? errorTxtMaxLines;
  final Widget? iconImage;
  final Color? iconImageColor;
  final bool? dense;
  final bool? collapsed;
  final String? counterTxt;
  final TextStyle? counterTxtStyle;
  final Widget? counterWidget;
  final String? semanticCounterTxt;
  final EdgeInsetsGeometry? contentEdgePadding;

  const InputDecorationTextField({
    this.counterTxt,
    this.counterTxtStyle,
    this.counterWidget,
    this.semanticCounterTxt,
    this.contentEdgePadding,
    this.helperTxt,
    this.helperTxtStyle,
    this.helperTxtMaxLines,
    this.onHoverColor,
    this.onFocusColor,
    this.hintTxtDirection,
    this.hintTxtStyle,
    this.hintTxtMaxLines,
    this.errorTxtStyle,
    this.errorTxtMaxLines,
    this.iconImage,
    this.iconImageColor,
    this.dense,
    this.collapsed,
    this.inputBorder,
    this.prefixTxt,
    this.suffixTxt,
    this.prefixIconImage,
    this.suffixIconImage,
    this.prefixConstraints,
    this.suffixConstraints,
    this.suffixColor,
    this.prefixColor,
    this.focusedErrorInputBorder,
    this.hint,
    this.errorMsg,
    this.isFilled,
    this.isEnabled,
    this.focusedInputBorder,
    this.errorInputBorder,
    this.enabledInputBorder,
    this.disabledInputBorder,
    this.filledColor,
    this.prefixTextStyling,
    this.suffixTextStyling,
  });

  @override
  Widget? get counter => counterWidget;

  @override
  TextStyle? get counterStyle => counterTxtStyle;

  @override
  String? get counterText => counterTxt;

  @override
  String? get semanticCounterText => semanticCounterTxt;

  @override
  EdgeInsetsGeometry? get contentPadding =>
      contentEdgePadding ?? const EdgeInsets.fromLTRB(12, 12, 12, 12);

  @override
  InputBorder? get focusedBorder => focusedInputBorder;

  @override
  bool get enabled => isEnabled ?? true;

  @override
  InputBorder? get enabledBorder => enabledInputBorder;

  @override
  InputBorder? get focusedErrorBorder => focusedErrorInputBorder;

  @override
  InputBorder? get disabledBorder => disabledInputBorder;

  @override
  bool? get filled => isFilled;

  @override
  Color? get fillColor => filledColor;

  @override
  InputBorder? get errorBorder => errorInputBorder;

  @override
  String? get hintText => hint;

  @override
  String? get errorText => errorMsg;

  @override
  Widget? get suffixIcon => suffixIconImage;

  @override
  Color? get suffixIconColor => suffixColor;

  @override
  TextStyle? get suffixStyle => suffixTextStyling;

  @override
  String? get suffixText => suffixTxt;

  @override
  BoxConstraints? get suffixIconConstraints => suffixConstraints;

  @override
  Widget? get prefixIcon => prefixIconImage;

  @override
  Color? get prefixIconColor => prefixColor;

  @override
  TextStyle? get prefixStyle => prefixTextStyling;

  @override
  String? get prefixText => prefixTxt;

  @override
  BoxConstraints? get prefixIconConstraints => prefixConstraints;

  @override
  int? get helperMaxLines => helperTxtMaxLines;

  @override
  TextStyle? get helperStyle => helperTxtStyle;

  @override
  String? get helperText => helperTxt;

  @override
  int? get hintMaxLines => hintTxtMaxLines;

  @override
  TextStyle? get hintStyle => hintTxtStyle;

  @override
  TextDirection? get hintTextDirection => hintTxtDirection;

  @override
  Color? get hoverColor => onHoverColor;

  @override
  Color? get focusColor => onFocusColor;

  @override
  int? get errorMaxLines => errorTxtMaxLines;

  @override
  TextStyle? get errorStyle => errorTxtStyle;

  @override
  Widget? get icon => iconImage;

  @override
  Color? get iconColor => iconImageColor;

  @override
  bool get isCollapsed => collapsed ?? true;

  @override
  bool? get isDense => dense;

  @override
  InputBorder? get border => inputBorder;
}


lib/webkul_textfield_with_label.dart
library textfield_with_label;

import 'dart:core';
import 'dart:ui';
import 'package:flutter/material.dart';
import 'package:webkul_textfield_with_label/widget/textfield.dart';
import 'utils/input_decoration.dart';

class TextFieldWithLabel extends StatefulWidget {
  final double? padding;
  final String labelText;
  final bool? isRequired;
  final String? isRequiredText;
  final TextStyle? isRequiredTextStyle;
  final TextStyle? labelTextStyle;
  final double? gapBtwLblAndField;
  final TextEditingController? controller;
  final String? initialValue;
  final FocusNode? focusNode;
  final bool? autoCorrect;
  final bool? autoFocus;
  final AutovalidateMode? autoValidateMode;
  final bool? readOnly;
  final int? maxLines;
  final bool? obscureText;
  final Radius? cursorRadius;
  final VoidCallback? onTap;
  final bool? enabled;
  final double? cursorHeight;
  final Color? cursorColor;
  final String? obscureCharacter;
  final TextAlign? textAlign;
  final TextStyle? textFieldTextStyle;
  final ValueChanged<String>? onChanged;
  final String? Function(String?)? validator;
  final VoidCallback? onEditingComplete;
  final ValueChanged<String?>? onSaved;
  final int? minLines;
  final Iterable<String>? autoFillHints;
  final Brightness? keyboardAppearance;
  final InputCounterWidgetBuilder? inputCounterWidgetBuilder;
  final double? cursorWidth;
  final TextInputType? keyboardType;
  final TextInputAction? textInputAction;
  final TextCapitalization? textCapitalization;
  final bool? showCursor;
  final ValueChanged<String>? onFieldSubmitted;
  final TextDirection? textDirection;
  final InputDecorationTextField? inputDecorationTextField;

  const TextFieldWithLabel(
      {Key? key,
      required this.labelText,
      this.inputDecorationTextField,
      this.padding,
      this.labelTextStyle,
      this.gapBtwLblAndField,
      this.controller,
      this.initialValue,
      this.focusNode,
      this.autoCorrect,
      this.maxLines,
      this.textAlign,
      this.textFieldTextStyle,
      this.readOnly,
      this.onChanged,
      this.onTap,
      this.onSaved,
      this.obscureText,
      this.obscureCharacter,
      this.autoFocus,
      this.autoValidateMode,
      this.validator,
      this.enabled,
      this.cursorHeight,
      this.cursorColor,
      this.onEditingComplete,
      this.cursorRadius,
      this.minLines,
      this.autoFillHints,
      this.keyboardAppearance,
      this.cursorWidth,
      this.inputCounterWidgetBuilder,
      this.keyboardType,
      this.textInputAction,
      this.textCapitalization,
      this.showCursor,
      this.onFieldSubmitted,
      this.textDirection,
      this.isRequired,
      this.isRequiredText,
      this.isRequiredTextStyle})
      : super(key: key);

  @override
  State<TextFieldWithLabel> createState() => _TextFieldWithLabelState();
}

class _TextFieldWithLabelState extends State<TextFieldWithLabel> {
  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.all(widget.padding ?? 8.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          RichText(
              text: TextSpan(children: [
            TextSpan(
              text: widget.labelText,
              style: widget.labelTextStyle ??
                  const TextStyle(color: Colors.black, fontSize: 16.0),
            ),
            widget.isRequired == true
                ? WidgetSpan(
                    child: Text(widget.isRequiredText ?? "*",
                        style: widget.isRequiredTextStyle ??
                            const TextStyle(
                              color: Colors.red,
                              fontSize: 16.0,
                              fontFeatures: <FontFeature>[
                                FontFeature.superscripts(),
                              ],
                            )),
                  )
                : const TextSpan(text: ""),
          ])),
          SizedBox(
            height: widget.gapBtwLblAndField ?? 6.0,
          ),
          LabelTextFormField(
              key: widget.key,
              showCursor: widget.showCursor ?? true,
              textDirection: widget.textDirection,
              textCapitalization:
                  widget.textCapitalization ?? TextCapitalization.none,
              textInputAction: widget.textInputAction,
              onFieldSubmitted: widget.onFieldSubmitted,
              keyboardAppearance: widget.keyboardAppearance,
              autofillHints: widget.autoFillHints,
              buildCounter: widget.inputCounterWidgetBuilder,
              cursorRadius: widget.cursorRadius,
              cursorWidth: widget.cursorWidth ?? 2.0,
              keyboardType: widget.keyboardType,
              onTap: widget.onTap,
              minLines: widget.minLines,
              cursorColor: widget.cursorColor,
              cursorHeight: widget.cursorHeight,
              enabled: widget.enabled,
              focusNode: widget.focusNode,
              onEditingComplete: widget.onEditingComplete,
              onSaved: widget.onSaved,
              validator: widget.validator,
              autofocus: widget.autoFocus ?? false,
              autovalidateMode: widget.autoValidateMode,
              readOnly: widget.readOnly ?? false,
              onChanged: widget.onChanged,
              obscureText: widget.obscureText ?? false,
              obscuringCharacter: widget.obscureCharacter ?? "*",
              initialValue: widget.initialValue,
              textAlign: widget.textAlign ?? TextAlign.start,
              style: widget.textFieldTextStyle,
              maxLines: widget.maxLines,
              autocorrect: widget.autoCorrect ?? false,
              controller: widget.controller,
              decoration: widget.inputDecorationTextField)
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:webkul_textfield_with_label/utils/input_decoration.dart';
import 'package:webkul_textfield_with_label/webkul_textfield_with_label.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  TextEditingController? emailController;
  TextEditingController? pwdController;
  bool isPasswordVisible = false;
  final _formKey = GlobalKey<FormState>();

  @override
  void initState() {
    emailController = TextEditingController();
    pwdController = TextEditingController();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.grey.shade200,
      body: Center(
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: SingleChildScrollView(
            child: Form(
              key: _formKey,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  const Text(
                    "Hello Again !!",
                    style:
                        TextStyle(fontSize: 24.0, fontWeight: FontWeight.bold),
                  ),
                  const SizedBox(
                    height: 12.0,
                  ),
                  const Text(
                    "Welcome back you've been missed!!",
                    style:
                        TextStyle(fontSize: 20.0, fontWeight: FontWeight.bold),
                  ),
                  const SizedBox(
                    height: 24.0,
                  ),
                  TextFieldWithLabel(
                      controller: emailController,
                      labelText: "Email",
                      isRequired: true,
                      inputDecorationTextField: InputDecorationTextField(
                        prefixIconImage: const Icon(Icons.person),
                        hint: 'Enter Username',
                        filledColor: Colors.white,
                        isEnabled: true,
                        isFilled: true,
                        contentEdgePadding: const EdgeInsets.all(14.0),
                        focusedInputBorder: OutlineInputBorder(
                            borderSide: const BorderSide(color: Colors.white),
                            borderRadius: BorderRadius.circular(8.0)),
                        enabledInputBorder: OutlineInputBorder(
                            borderSide: const BorderSide(color: Colors.white),
                            borderRadius: BorderRadius.circular(8.0)),
                        errorInputBorder: OutlineInputBorder(
                            borderSide: const BorderSide(color: Colors.red),
                            borderRadius: BorderRadius.circular(8.0)),
                      ),
                      validator: (value) {
                        if (!RegExp(
                                r"^[a-zA-Z0-9.a-zA-Z0-9.!#$%&'*+-/=?^_`{|}~]+@[a-zA-Z0-9]+\.[a-zA-Z]+")
                            .hasMatch(value ?? "")) {
                          return "Incorrect Email";
                        }
                        return null;
                      }),
                  TextFieldWithLabel(
                    controller: pwdController,
                    isRequired: true,
                    obscureText: isPasswordVisible ? false : true,
                    maxLines: 1,
                    labelText: 'Password',
                    inputDecorationTextField: InputDecorationTextField(
                      prefixIconImage: const Icon(Icons.lock),
                      isFilled: true,
                      hint: 'Enter Password',
                      filledColor: Colors.white,
                      isEnabled: true,
                      suffixIconImage: GestureDetector(
                          onTap: () {
                            setState(() {
                              isPasswordVisible = !isPasswordVisible;
                            });
                          },
                          child: Icon(isPasswordVisible
                              ? Icons.visibility
                              : Icons.visibility_off)),
                      contentEdgePadding: const EdgeInsets.all(14.0),
                      focusedInputBorder: OutlineInputBorder(
                          borderSide: const BorderSide(color: Colors.white),
                          borderRadius: BorderRadius.circular(8.0)),
                      enabledInputBorder: OutlineInputBorder(
                          borderSide: const BorderSide(color: Colors.white),
                          borderRadius: BorderRadius.circular(8.0)),
                      errorInputBorder: OutlineInputBorder(
                          borderSide: const BorderSide(color: Colors.red),
                          borderRadius: BorderRadius.circular(8.0)),
                    ),
                    validator: (value) {
                      if (RegExp(
                              r'^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[!@#$&*~]).{8,}$')
                          .hasMatch(value ?? "")) {
                        return "Please add a Strong Password";
                      }
                      return null;
                    },
                  ),
                  const SizedBox(
                    height: 24.0,
                  ),
                  Padding(
                    padding: const EdgeInsets.all(8.0),
                    child: SizedBox(
                        height: 50.0,
                        width: MediaQuery.of(context).size.width,
                        child: ElevatedButton(
                            style: ElevatedButton.styleFrom(
                              backgroundColor: Colors.pink,
                            ),
                            onPressed: () {
                              if (_formKey.currentState!.validate()) {
                                _formKey.currentState!.save();
                                debugPrint("Email-->${emailController?.text}");
                              }
                            },
                            child: const Text("Login"))),
                  )
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
